### PR TITLE
[Feature/Identity] Persistence for the Identity module using a System Index

### PR DESCRIPTION
**Pull request by cwperks**
### Description

This PR introduces a system index for the identity module to persist identity config like internal users to a system index in OpenSearch. The default index is `.identity_config` on this PR and sources the index on bootstrap if it does not exist. This PR also introduces the concept of DynamicConfig where listeners can be registered to react to changes in the config which will be useful when creating APIs to add, change and remove users. 

The bootstrap process works by overriding `onNodeStarted` and calling `ConfigurationRepository.initOnNodeStarted()` which then reads in the config files, currently only `internal_users.yml`, and populates the `.identity_config` index. 

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.common.xcontent.XContentParser;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.identity.ConfigConstants;
+import org.opensearch.identity.utils.SecurityUtils;
+import org.opensearch.threadpool.ThreadPool;
+
+import static org.opensearch.common.xcontent.DeprecationHandler.THROW_UNSUPPORTED_OPERATION;
+
+public class ConfigurationLoader {
```
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java:47]: Can you add a brief javadoc of what this class does/ why is it needed?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.MappingMetadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.util.concurrent.ThreadContext.StoredContext;
+import org.opensearch.env.Environment;
+import org.opensearch.threadpool.ThreadPool;
+
+public class ConfigurationRepository {
```
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:49]: nit: Javadoc for this class

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    private final AtomicBoolean installDefaultConfig = new AtomicBoolean();
+
+    private ConfigurationRepository(
+        Settings settings,
+        final Path configPath,
+        ThreadPool threadPool,
+        Client client,
+        ClusterService clusterService
+    ) {
+        this.securityIndex = settings.get(ConfigConstants.IDENTITY_CONFIG_INDEX_NAME, ConfigConstants.IDENTITY_DEFAULT_CONFIG_INDEX);
```
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:72]: nit: we can rename this to identityIndex throughout this file

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+                    LOGGER.info("Background init thread started. Install default config?: " + installDefaultConfig.get());
+
+                    if (installDefaultConfig.get()) {
+
+                        try {
+                            String lookupDir = System.getProperty("identity.default_init.dir");
+                            final String cd = lookupDir != null
+                                ? (lookupDir + "/")
+                                : new Environment(settings, configPath).configDir().toAbsolutePath().toString() + "/";
+                            Path confFile = Path.of(cd + "internal_users.yml");
```
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:93]: this is a static list that will be loaded on server start, correct?
If so, does it mean when this config is uploaded the existing user indices, if any, will be deleted?
I saw the implementation of getConfigurationFromIndex but was still confused

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java [hunk 5]
```diff
@@ -43,11 +52,172 @@ public static void inject(final InjectableValues.Std injectableValues) {
...
+                    return (omitDefaults ? defaulOmittingObjectMapper : objectMapper).writeValueAsString(value);
+                }
+            });
+        } catch (final PrivilegedActionException e) {
+            throw (JsonProcessingException) e.getCause();
+        }
+
+    }
+
+    public static <T> T readValue(String string, TypeReference<T> tr) throws IOException {
```
Comment (DarshitChanpura) [inline sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java:170]: For my knowledge: why do we need this readValue and one one line 189, and in which scenarios will they be used?

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java [hunk 6]
```diff
@@ -43,11 +52,172 @@ public static void inject(final InjectableValues.Std injectableValues) {
...
+        } catch (final PrivilegedActionException e) {
+            throw (IOException) e.getCause();
+        }
+    }
+
+    public static TypeFactory getTypeFactory() {
+        return objectMapper.getTypeFactory();
+    }
+
+    public static Set<String> getFields(Class<?> cls) {
```
Comment (DarshitChanpura) [inline sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java:215]: For my knowlege, when will this method be used

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java [hunk 7]
```diff
@@ -18,23 +18,23 @@
...
     @JsonProperty(value = "hash")
     private String bcryptHash;
 
     @JsonProperty(value = "attributes")
     private Map<String, String> attributes = Collections.emptyMap();
 
-    @JsonProperty(value = "primary_principal")
-    public StringPrincipal getPrimaryPrincipal() {
-        return primaryPrincipal;
+    @JsonProperty(value = "username")
```
Comment (DarshitChanpura) [inline sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:30]: Why this change, are we renaming Principal terminology or are we reserving it for something else?

**Review by DarshitChanpura — COMMENTED** [review]
Thank you @cwperks for adding this base for persistence!

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.configuration;
+
+public interface Initializable {
```
Comment (peternied) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java:11]: Why make this interface?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/DynamicConfigFactory.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.threadpool.ThreadPool;
+import org.greenrobot.eventbus.EventBus;
+import org.greenrobot.eventbus.EventBusBuilder;
+
+public class DynamicConfigFactory implements Initializable, ConfigurationChangeListener {
+
+    public static final EventBusBuilder EVENT_BUS_BUILDER = EventBus.builder();
```
Comment (peternied) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/DynamicConfigFactory.java:27]: I like the usage of event bus for this rather than rolling our own, but that said have you seen how this is done elsewhere in OpenSearch?

**Review by peternied — CHANGES_REQUESTED** [review]
Nice job taking this on!  I think I've got some design related questions, maybe a quick mermaid diagram of what is interacting where when would help improve the tradeoffs you were making.

I see a lot of security plugin code making its way into this change that might be better to minimize.

Finally, is there a reason we can't reuse the existing settings configuration?  that way there is only a signal standard.  Note; it would be appropriate for the bootstrap workflow to be seprate from the index usage

File: sandbox/modules/identity/build.gradle [hunk 10]
```diff
@@ -20,6 +20,9 @@ dependencies {
   api project(':libs:opensearch-core')
   api project(':sandbox:libs:opensearch-authn')
 
+  implementation 'org.greenrobot:eventbus:3.2.0'
+  implementation 'commons-cli:commons-cli:1.3.1'
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/build.gradle:24]: commons-cli can be removed - in a previous iteration I also had the `main` method inside Hasher, but that was removed along with any reference to anything in commons-cli

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/Hasher.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.Objects;
+
+public class Hasher {
+
+    public static String hash(final char[] clearTextPassword) {
+        final byte[] salt = new byte[16];
+        new SecureRandom().nextBytes(salt);
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/Hasher.java:21]: I'll add a TODO statement in here, we currently only support BCrypt hashing but may want to support other hashing methods in the future.

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java [hunk 12]
```diff
@@ -43,11 +52,172 @@ public static void inject(final InjectableValues.Std injectableValues) {
         defaulOmittingObjectMapper.setInjectableValues(injectableValues);
     }
 
+    public static boolean getOrDefault(Map<String, Object> properties, String key, boolean defaultValue) throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java:55]: I'll remove the unused methods from this class, they can be added in later as needed.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/SecurityUtils.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+import org.opensearch.common.settings.Settings;
+import org.opensearch.identity.ConfigConstants;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class SecurityUtils {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/utils/SecurityUtils.java:22]: Removing this class

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/SecurityDynamicConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.common.xcontent.XContentType;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class SecurityDynamicConfiguration<T> implements ToXContent {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/SecurityDynamicConfiguration.java:29]: Yes, this scaffolding allows us to dynamically react to changes in the system index. If its not added in this PR, then it will need to be as soon as we create an API like creating a new user.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java (hunk 8 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java:11]: Do you think it should be removed? A lot of this is modeled after the security plug-in with a lot of extraneous code removed out and core functionality kept in. This is only ever implemented once, I guess if there are multiple implementers then it would be more compelling to have an interface.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/DynamicConfigFactory.java (hunk 9 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/DynamicConfigFactory.java:27]: Do you know of examples within core of doing async programming to listen for changes on an index? Possibly `IndexOperationListener`?

One of the challenges is making sure that nodes are synchronized and the method introduced here reloads the index into memory every 3s using the event bus.

Comment (peternied) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/DynamicConfigFactory.java:27]: @saratvemulapalli It looks like you fielding ad-hoc requests this week - did you have any recommendations for this sub/pub processes in OpenSearch?

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java (hunk 7 above)
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:30]: I think we should use `username` for the primary principal as a convention. Do you think this should be changed back? I think using username makes the work in identity more readable as username is currently the primary identifier for a subject in OpenSearch.

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java (hunk 6 above)
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java]: I removed the unused methods of this class. They can be added back in as needed for future PRs.

File: sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/UserPersistenceIT.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        System.setProperty("identity.default_init.dir", defaultInitDirectory);
+
+        startNodes();
+
+        ClusterHealthResponse clusterHealthResponse = client().admin().cluster().prepareHealth().setClusterManagerNodeTimeout("1s").get();
+
+        assertTrue(
+            ConfigConstants.IDENTITY_DEFAULT_CONFIG_INDEX + " index exists",
+            clusterHealthResponse.getIndices().containsKey(ConfigConstants.IDENTITY_DEFAULT_CONFIG_INDEX)
+        );
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/UserPersistenceIT.java:81]: Done

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java (hunk 5 above)
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java:103]: `public static <T> T readValue(String string, Class<T> clazz)` is used to read from the `internal_users.yml` and map to User objects. All other readValue that weren't being used were removed.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java (hunk 4 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:93]: That's handled in `ConfigHelper.uploadFile`

```
try (Reader reader = createFileOrStringReader(cType, configVersion, filepath, populateEmptyIfFileMissing)) {

    final IndexRequest indexRequest = new IndexRequest(index).id(configType)
        .opType(OpType.CREATE)
        .setRefreshPolicy(RefreshPolicy.IMMEDIATE)
        .source(configType, readXContent(reader, XContentType.YAML));
    IndexResponse response = tc.index(indexRequest).actionGet();
    final String res = response.getId();

    if (!configType.equals(res)) {
        throw new Exception(
            "   FAIL: Configuration for '" + configType + "' failed for unknown reasons. Pls. consult logfile of opensearch"
        );
    }
    LOGGER.info("Doc with id '{}' and version {} is updated in {} index.", configType, configVersion, index);
} catch (VersionConflictEngineException versionConflictEngineException) {
    LOGGER.info("Index {} already contains doc with id {}, skipping update.", index, configType);
}
```

If it exists, then a `VersionConflictEngineException` is thrown and handled by skipping.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java (hunk 3 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:72]: Good idea

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java (hunk 1 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java:49]: Added, this class loads the identity configuration from the system index into memory periodically.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java (hunk 2 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java:52]: Added javadoc here as well. This class loads the identity configuration from files into a system index. It initializes the system index if it does not exist.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java (hunk 3 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationRepository.java]: Done

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+    private final Client client;
+    private final String securityIndex;
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/ConfigurationLoader.java:53]: Changed

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java [hunk 17]
```diff
@@ -43,11 +52,172 @@ public static void inject(final InjectableValues.Std injectableValues) {
...
+
+    @SuppressWarnings("unchecked")
+    public static <T> T getOrDefault(Map<String, Object> properties, String key, T defaultValue) {
+        T value = (T) properties.get(key);
+        return value != null ? value : defaultValue;
+    }
+
+    public static <T> T readTree(JsonNode node, Class<T> clazz) throws IOException {
+
+        final SecurityManager sm = System.getSecurityManager();
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/DefaultObjectMapper.java]: Agreed. This class also uses SpecialPermission which I am not quite sure of its usage.

Here's an example snippet:

```
final SecurityManager sm = System.getSecurityManager();

if (sm != null) {
    sm.checkPermission(new SpecialPermission());
}
```

SpecialPermission is coming from OpenSearch here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/SpecialPermission.java

File: sandbox/modules/identity/build.gradle (hunk 10 above)
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/build.gradle]: Switched to eventbus-java:3.3.1

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java (hunk 8 above)
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/configuration/Initializable.java]: I removed this interface which was only used once.

File: sandbox/modules/identity/build.gradle (hunk 10 above)
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/build.gradle]: I switched back to eventbus:3.2.0. Switching to eventbus-java:3.3.1 brought in android files and I cannot figure out how to exclude them.

```
Unnecessary exclusions, following classes are not missing:
   * android.app.Activity
  * android.app.AlertDialog$Builder
  * android.app.Dialog
  * android.app.DialogFragment
  * android.app.Fragment
  * android.app.FragmentManager
  * android.app.FragmentTransaction
  * android.content.Context
  * android.content.DialogInterface
  * android.content.DialogInterface$OnClickListener
  * android.content.res.Resources
  * android.os.Build$VERSION
  * android.os.Bundle
  * android.os.Handler
  * android.os.Looper
  * android.os.Message
  * android.os.SystemClock
  * android.support.v4.app.DialogFragment
  * android.support.v4.app.Fragment
  * android.support.v4.app.FragmentActivity
  * android.support.v4.app.FragmentManager
  * android.support.v4.app.FragmentTransaction
  * android.util.Log
```

Is there a way to exclude pulling in these files when adding a dependency on eventbus-java:3.3.1 if we use eventbus for async programming?

File: sandbox/modules/identity/src/internalClusterTest/resources/invalid/internal_users.yml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+# Invalid internal_users.yml, hash is required
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/internalClusterTest/resources/invalid/internal_users.yml:1]: This is used by `UserPersistenceIT` to test the reading of the file on startup. I'll look into modifying the test to keep this in line.

**Comment by cwperks** [comment]
Thank you @peternied. I addressed many of the comments, but there are still a few remaining like creating a diagram. The next couple of PRs to follow will be on APIs for interacting with the internal IdP to create, list, update and delete users along with getting information about the current user (whoami). This PR presents a base for persisting data that will be used while interacting with the IdP.

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java (hunk 7 above)
Comment (DarshitChanpura) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/User.java:30]: it should work for now, but this should definitely be decided upon as to what naming are we proceeding with

**Review by DarshitChanpura — APPROVED** [review]
LGTM. I'm using it as a base for creating APIs, and so thank you @cwperks for this PR!

File: sandbox/modules/identity/src/internalClusterTest/resources/invalid/internal_users.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/src/internalClusterTest/resources/invalid/internal_users.yml]: These are now being created inline using temporary files

**Comment by cwperks** [comment]
The gradle check failure is also happening on `feature/identity` too:

```
> Task :distribution:bwc:minor:buildBwcLinuxTar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 2.5.0 didn't generate expected file /Users/cwperx/Projects/opensearch/OpenSearch/distribution/bwc/minor/build/bwc/checkout-2.x/distribution/archives/linux-tar/build/distributions/opensearch-min-2.5.0-SNAPSHOT-linux-x64.tar.gz

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 28s
18 actionable tasks: 12 executed, 6 up-to-date
➜  openSearch git:(feature/identity) ls /Users/cwperx/Projects/opensearch/OpenSearch/distribution/bwc/minor/build/bwc/checkout-2.x/distribution/archives/linux-tar/build/distributions
opensearch-min-2.6.0-SNAPSHOT-linux-x64.tar.gz
```

**Comment by cwperks** [comment]
The build on main is broken with the same error:

```
Execution failed for task ':qa:rolling-upgrade:v2.6.0#oneThirdUpgradedTest'.
```

**Comment by cwperks** [comment]
@peternied I updated this PR with documentation on the initialization process and the update process for the identity system index.

**Comment by cwperks** [comment]
Looks like its not displaying again and I'm not sure why. I'll try re-uploading. Here is the diagram:

identity_configuration_bootstrap_and_update

**Comment by peternied** [comment]
@cwperks Please convert them to mermaid format so they can be edited by other users, here is a quick conversion of the chart you added

```mermaid
flowchart TD;
  bs[Bootstrap]-->ons[IdentityPlugin.onNodeStarted];
  ons-->ions[ConfigurationRepository.initOnNodeStart];
  ions-->bg[Background thread reads config files<br>internal_users.yml,<br> stores in system index];
  bg-->imi[Identity Module is initialized];
```

```
flowchart TD;
  bs[Bootstrap]-->ons[IdentityPlugin.onNodeStarted];
  ons-->ions[ConfigurationRepository.initOnNodeStart];
  ions-->bg[Background thread reads config files<br>internal_users.yml,<br> stores in system index];
  bg-->imi[Identity Module is initialized];
```

**Comment by cwperks** [comment]
@peternied The diagrams are now mermaid diagrams.

**Review by peternied — APPROVED** [review]

File: sandbox/modules/identity/build.gradle [hunk 19]
```diff
@@ -21,6 +21,8 @@ dependencies {
   api project(':libs:opensearch-core')
   implementation project(':sandbox:libs:opensearch-authn')
 
+  implementation 'org.greenrobot:eventbus:3.2.0'
```
[the remark this answers is not in the document]
Comment (cwperks) [reply sandbox/modules/identity/build.gradle:24]: See the comment here: https://github.com/opensearch-project/OpenSearch/pull/5613#discussion_r1063771212

I switched to eventbus-java, but it started bringing in extraneous classes that I was not able to figure out how to exclude.
